File: src/client/python/agentguard/schemas/events.py
"""RuntimeEvent: normalized representation of any runtime behavior."""
from __future__ import annotations

import re
import uuid
from dataclasses import dataclass, field
from enum import Enum
from typing import Any

from agentguard.schemas.context import RuntimeContext
from agentguard.utils.hash import stable_hash
from agentguard.utils.time import now_ts


class EventType(str, Enum):
    LLM_INPUT = "llm_input"
    LLM_OUTPUT = "llm_output"
    TOOL_INVOKE = "tool_invoke"
    TOOL_RESULT = "tool_result"

    # Deprecated event types intentionally kept out of the active enum:
    # user_input, llm_thought, llm_tool_call_candidate, memory_read,
    # memory_write, file_read, file_write, network_request, final_response,
    # sandbox_execution, policy_decision.


# Patterns used for redaction of sensitive payload values.
_SECRET_KEY_HINTS = (
    "password",
    "passwd",
    "secret",
    "token",
    "api_key",
    "apikey",
    "authorization",
    "access_key",
    "private_key",
)
_REDACT_PATTERNS = [
    re.compile(r"sk-[A-Za-z0-9]{8,}"),
    re.compile(r"AKIA[0-9A-Z]{12,}"),
    re.compile(r"ghp_[A-Za-z0-9]{20,}"),
    re.compile(r"\b\d{13,19}\b"),  # card-like
]
_REDACTED = "[REDACTED]"


def _redact_value(value: Any, key: str | None = None) -> Any:
    if key and any(h in key.lower() for h in _SECRET_KEY_HINTS):
        return _REDACTED
    if isinstance(value, str):
        out = value
        for pat in _REDACT_PATTERNS:
            out = pat.sub(_REDACTED, out)
        return out
    if isinstance(value, dict):
        return {k: _redact_value(v, k) for k, v in value.items()}
    if isinstance(value, list):
        return [_redact_value(v) for v in value]
    return value


@dataclass
class RuntimeEvent:
    """A single normalized runtime event."""

    event_id: str
    event_type: EventType
    timestamp: float
    context: RuntimeContext
    payload: dict[str, Any]
    risk_signals: list[str] = field(default_factory=list)
    metadata: dict[str, Any] = field(default_factory=dict)

    # ---- serialization -------------------------------------------------
    def to_dict(self) -> dict[str, Any]:
        return {
            "event_id": self.event_id,
            "event_type": self.event_type.value,
            "timestamp": self.timestamp,
            "context": self.context.to_dict(),
            "payload": self.payload,
            "risk_signals": list(self.risk_signals),
            "metadata": self.metadata,
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> "RuntimeEvent":
        return cls(
            event_id=data.get("event_id") or _new_id(),
            event_type=EventType(data["event_type"]),
            timestamp=float(data.get("timestamp") or now_ts()),
            context=RuntimeContext.from_dict(data.get("context") or {}),
            payload=dict(data.get("payload") or {}),
            risk_signals=list(data.get("risk_signals") or []),
            metadata=dict(data.get("metadata") or {}),
        )

    def redacted(self) -> "RuntimeEvent":
        """Return a copy with secrets removed from payload/metadata."""
        return RuntimeEvent(
            event_id=self.event_id,
            event_type=self.event_type,
            timestamp=self.timestamp,
            context=self.context,
            payload=_redact_value(self.payload),
            risk_signals=list(self.risk_signals),
            metadata=_redact_value(self.metadata),
        )

    def stable_hash(self) -> str:
        """Deterministic hash ignoring volatile fields (id/timestamp)."""
        return stable_hash(
            {
                "event_type": self.event_type.value,
                "context": {
                    "session_id": self.context.session_id,
                    "policy": self.context.policy,
                    "policy_version": self.context.policy_version,
                },
                "payload": self.payload,
                "risk_signals": sorted(self.risk_signals),
            }
        )

    def add_signal(self, signal: str) -> None:
        if signal and signal not in self.risk_signals:
            self.risk_signals.append(signal)


def _new_id() -> str:
    return f"evt_{uuid.uuid4().hex[:16]}"


def _make(
    event_type: EventType,
    context: RuntimeContext,
    payload: dict[str, Any] | None = None,
    *,
    risk_signals: list[str] | None = None,
    metadata: dict[str, Any] | None = None,
) -> RuntimeEvent:
    return RuntimeEvent(
        event_id=_new_id(),
        event_type=event_type,
        timestamp=now_ts(),
        context=context,
        payload=payload or {},
        risk_signals=risk_signals or [],
        metadata=metadata or {},
    )


# ---- helper constructors ----------------------------------------------
def user_input(context: RuntimeContext, text: str, **meta: Any) -> RuntimeEvent:
    """Compatibility alias: user text is now represented as LLM_INPUT."""
    return _make(
        EventType.LLM_INPUT,
        context,
        {"text": text, "messages": [{"role": "user", "content": text}]},
        metadata=meta,
    )


def llm_input(context: RuntimeContext, messages: Any, **meta: Any) -> RuntimeEvent:
    return _make(EventType.LLM_INPUT, context, {"messages": messages}, metadata=meta)


def llm_output(context: RuntimeContext, output: Any, **meta: Any) -> RuntimeEvent:
    return _make(EventType.LLM_OUTPUT, context, {"output": output}, metadata=meta)


def llm_thought(context: RuntimeContext, thought: str, **meta: Any) -> RuntimeEvent:
    """Compatibility alias: thoughts are no longer a separate event type."""
    return _make(EventType.LLM_OUTPUT, context, {"output": thought}, metadata=meta)


def tool_invoke(
    context: RuntimeContext,
    tool_name: str,
    arguments: dict[str, Any],
    *,
    capabilities: list[str] | None = None,
    **meta: Any,
) -> RuntimeEvent:
    payload = {
        "tool_name": tool_name,
        "arguments": arguments,
        "capabilities": capabilities or [],
    }
    return _make(EventType.TOOL_INVOKE, context, payload, metadata=meta)


def tool_result(
    context: RuntimeContext,
    tool_name: str,
    result: Any,
    *,
    error: str | None = None,
    **meta: Any,
) -> RuntimeEvent:
    payload = {"tool_name": tool_name, "result": result, "error": error}
    return _make(EventType.TOOL_RESULT, context, payload, metadata=meta)


def final_response(context: RuntimeContext, text: str, **meta: Any) -> RuntimeEvent:
    """Compatibility alias: final text is now represented as LLM_OUTPUT."""
    return _make(EventType.LLM_OUTPUT, context, {"output": text}, metadata=meta)

File: src/client/js/agentguard/schemas/events.js
"use strict";

const crypto = require("crypto");
const { RuntimeContext } = require("./context");
const { stableHash } = require("../utils/hash");
const { nowTs } = require("../utils/time");

const EventType = Object.freeze({
  LLM_INPUT: "llm_input",
  LLM_OUTPUT: "llm_output",
  TOOL_INVOKE: "tool_invoke",
  TOOL_RESULT: "tool_result",
});

const SECRET_KEY_HINTS = [
  "password",
  "passwd",
  "secret",
  "token",
  "api_key",
  "apikey",
  "authorization",
  "access_key",
  "private_key",
];
const REDACT_PATTERNS = [/sk-[A-Za-z0-9]{8,}/g, /AKIA[0-9A-Z]{12,}/g, /ghp_[A-Za-z0-9]{20,}/g, /\b\d{13,19}\b/g];
const REDACTED = "[REDACTED]";

function redactValue(value, key = null) {
  if (key && SECRET_KEY_HINTS.some((hint) => key.toLowerCase().includes(hint))) {
    return REDACTED;
  }
  if (typeof value === "string") {
    return REDACT_PATTERNS.reduce((current, pattern) => current.replace(pattern, REDACTED), value);
  }
  if (Array.isArray(value)) {
    return value.map((item) => redactValue(item));
  }
  if (value && typeof value === "object") {
    return Object.fromEntries(Object.entries(value).map(([childKey, childValue]) => [childKey, redactValue(childValue, childKey)]));
  }
  return value;
}

class RuntimeEvent {
  constructor(data = {}) {
    this.event_id = data.event_id || data.eventId || newId();
    this.event_type = data.event_type || data.eventType;
    this.timestamp = Number(data.timestamp ?? nowTs());
    this.context = data.context instanceof RuntimeContext ? data.context : RuntimeContext.fromDict(data.context || {});
    this.payload = { ...(data.payload || {}) };
    this.risk_signals = [...(data.risk_signals || data.riskSignals || [])];
    this.metadata = { ...(data.metadata || {}) };
  }

  toDict() {
    return {
      event_id: this.event_id,
      event_type: this.event_type,
      timestamp: this.timestamp,
      context: this.context.toDict(),
      payload: this.payload,
      risk_signals: [...this.risk_signals],
      metadata: { ...this.metadata },
    };
  }

  redacted() {
    return new RuntimeEvent({
      ...this.toDict(),
      payload: redactValue(this.payload),
      metadata: redactValue(this.metadata),
    });
  }

  stableHash() {
    return stableHash({
      event_type: this.event_type,
      context: {
        session_id: this.context.session_id,
        policy: this.context.policy,
        policy_version: this.context.policy_version,
      },
      payload: this.payload,
      risk_signals: [...this.risk_signals].sort(),
    });
  }

  addSignal(signal) {
    if (signal && !this.risk_signals.includes(signal)) {
      this.risk_signals.push(signal);
    }
  }

  static fromDict(data = {}) {
    return new RuntimeEvent(data);
  }
}

function newId() {
  return `evt_${crypto.randomBytes(8).toString("hex")}`;
}

function makeEvent(eventType, context, payload = {}, options = {}) {
  return new RuntimeEvent({
    event_type: eventType,
    context,
    payload,
    metadata: options.metadata || options.meta || {},
    risk_signals: options.risk_signals || options.riskSignals || [],
  });
}

function user_input(context, text, meta = {}) {
  return makeEvent(EventType.LLM_INPUT, context, { text, messages: [{ role: "user", content: text }] }, { metadata: meta });
}

function llm_input(context, messages, meta = {}) {
  return makeEvent(EventType.LLM_INPUT, context, { messages }, { metadata: meta });
}

function llm_output(context, output, meta = {}) {
  return makeEvent(EventType.LLM_OUTPUT, context, { output }, { metadata: meta });
}

function llm_thought(context, thought, meta = {}) {
  return llm_output(context, thought, meta);
}

function tool_invoke(context, tool_name, arguments_, options = {}) {
  return makeEvent(
    EventType.TOOL_INVOKE,
    context,
    {
      tool_name,
      arguments: arguments_,
      capabilities: options.capabilities || [],
    },
    { metadata: options.meta || options.metadata || {} }
  );
}

function tool_result(context, tool_name, result, options = {}) {
  return makeEvent(
    EventType.TOOL_RESULT,
    context,
    {
      tool_name,
      result,
      error: options.error || null,
    },
    { metadata: options.meta || options.metadata || {} }
  );
}

function final_response(context, text, meta = {}) {
  return llm_output(context, text, meta);
}

module.exports = {
  EventType,
  RuntimeEvent,
  user_input,
  llm_input,
  llm_output,
  llm_thought,
  tool_invoke,
  tool_result,
  final_response,
};
